File: android/src/main/kotlin/com/pravera/flutter_foreground_task/models/ForegroundTaskEventAction.kt
package com.pravera.flutter_foreground_task.models

import org.json.JSONObject
import java.util.Objects

data class ForegroundTaskEventAction(
    val type: ForegroundTaskEventType,
    val interval: Long
) {
    companion object {
        private const val TASK_EVENT_TYPE_KEY = "taskEventType"
        private const val TASK_EVENT_INTERVAL_KEY = "taskEventInterval"

        fun fromJsonString(jsonString: String): ForegroundTaskEventAction {
            val jsonObj = JSONObject(jsonString)

            val type: ForegroundTaskEventType = if (jsonObj.isNull(TASK_EVENT_TYPE_KEY)) {
                ForegroundTaskEventType.NOTHING
            } else {
                val value = jsonObj.getInt(TASK_EVENT_TYPE_KEY)
                ForegroundTaskEventType.fromValue(value)
            }

            val interval: Long = if (jsonObj.isNull(TASK_EVENT_INTERVAL_KEY)) {
                5000L
            } else {
                val value = jsonObj.getInt(TASK_EVENT_INTERVAL_KEY)
                value.toLong()
            }

            return ForegroundTaskEventAction(type = type, interval = interval)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (other == null || other !is ForegroundTaskEventAction) {
            return false
        }
        return this.type.value == other.type.value && this.interval == other.interval
    }

    override fun hashCode(): Int {
        return Objects.hash(type.value, interval)
    }
}

File: ios/Classes/models/ForegroundTaskEventAction.swift
//
//  ForegroundTaskEventAction.swift
//  flutter_foreground_task
//
//  Created by Woo Jin Hwang on 8/28/24.
//

import Foundation

private let TASK_EVENT_TYPE_KEY = "taskEventType"
private let TASK_EVENT_INTERVAL_KEY = "taskEventInterval"

struct ForegroundTaskEventAction: Equatable {
  let type: ForegroundTaskEventType
  let interval: Int
  
  static func fromJsonString(_ jsonString: String) -> ForegroundTaskEventAction {
    var type: ForegroundTaskEventType = .NOTHING
    var interval: Int = 5000
    
    if let jsonData = jsonString.data(using: .utf8) {
      if let jsonObj = try? JSONSerialization.jsonObject(with: jsonData, options: []) as? Dictionary<String, Any> {
        if let _type = jsonObj[TASK_EVENT_TYPE_KEY] as? Int {
          type = ForegroundTaskEventType.fromValue(_type)
        }
        
        if let _interval = jsonObj[TASK_EVENT_INTERVAL_KEY] as? Int {
          interval = _interval
        }
      }
    }
    
    return ForegroundTaskEventAction(type: type, interval: interval)
  }
  
  static func == (lhs: Self, rhs: Self) -> Bool {
    return lhs.type.rawValue == rhs.type.rawValue && lhs.interval == rhs.interval
  }
}
